### PR TITLE
Introduce temp watch protection for the bed

### PR DESCRIPTION
**Pull request by scalez**
When setting the bed temp via `M140`/`M190`, if the thermistor does not read an increase of `WATCH_BED_TEMP_INCREASE` degrees within `WATCH_BED_TEMP_PERIOD` seconds then it will throw "`Error:Heating failed, system stopped! Heater_ID: bed`" and call the `kill()` function.

### CODE REVIEW

**Comment by AnHardt** [comment]
This looks like a well made reinvention of my #2310 and it's followers.
If we don't have a maximum test temperature here, we will get a lot of false positives 
(https://github.com/MarlinFirmware/Marlin/pull/2310#issuecomment-116054413)

To find a consensus about the activation of this feature will be much easier these days.

Most of the beds out there do not need this feature at all, because they are self limiting and there is no risk of fire. But some (me) desperately need it for their monster-bed-heaters:
picture 162
picture 163

If merged in the RC-phase i recommend a deactivated by default version.

**Comment by scalez** [comment]
@AnHardt Not sure how I missed the fact that `millis_t` is typedef'ed as an `unsigned long` but now `WATCH_BED_TEMP_PERIOD` should be able to hold a value up to 4294967 seconds (49.71 days)! I set  `WATCH_BED_TEMP_PERIOD` to 60 seconds, I'd say that there should be a point where we can claim that if a bed does not reach its target temp within an extreme `WATCH_BED_TEMP_PERIOD` then the design is inherently flawed and the designer/user should be aware that what they're doing is potentially dangerous.

**Comment by jbrazio** [comment]
@AnHardt will you validate this PR with your ~1KW setup ? :-)

**Comment by AnHardt** [comment]
@jbrazio 
Sorry. No. No in the near future (days - weeks). Since Easter I'm still barely healthy enough to keep my business going and to drop a few lines now and then.

But as long the algorithm is the same as in the nozzle heater version, i have unlimited trust in it.
@scalez 
Alone finding reasonable default values for the case the target temperature is above the self-limiting-temperature, is an impossible task. If the bed will not reach the target temperature the user will realize that - it does not need a error message and a safety shutdown then and more important: "There is no risk of fire" . For that a maximum test temperature is needed and that has to be well below the self-limiting-temperature - if there is a self-limiting-temperature. If the is no self-limiting before self destruction -  the maximum test temperature can be MAX_TEMP_BED.).

**Comment by scalez** [comment]
@AnHardt sincerely hope you get better.

The logic is identical to the nozzle.
If the designer/user is aware of the self-limiting-temperature then the case when the target is unreachable should never happen. If the designer/user is setting the target temp using M140 and printing prior to the target temperature being reach (i.e. using M190) then that can also be considered a design/user error.

**Comment by AnHardt** [comment]
In the days when i begun to experiment with, heat activated epoxy's, on high temperature beds, before i got the power-beds, i used to say "give me all you have", give me 150°, even if the bed barely could reach 225° maybe 130° with closed printer doors.
If i remember correct, for an original K8200, "give me all" is the standard for printing ABS.

Of course this is not what you want, but the best you can get, and more common than you might think.

**Comment by thinkyhead** [comment]
This feature and the Bed Temp Residency checking have both been hovering on the periphery as features "we never wanted to add" because heated beds operate in a lower temp range and are –as pointed out– pretty well self-limiting. Still, for symmetry I don't think it's bad to have these two features added now. But I'm sure we can leave them disabled by default.

We ought to be leaving these extra bed temperature features (and a couple of other enhancements disabled by default) anyway, to accommodate the smaller (64K) boards that are now getting crowded out by the ballooning size of the Marlin binary.

With that in mind, keep an eye out for any chance to reduce the size of the "standard" or "default" install of Marlin. Aim for a target binary as small as version 1.0.2-1, or even smaller if possible.

**Comment by jbrazio** [comment]
There is no way to only disable this feature, it is enabled by `THERMAL_PROTECTION_BED` which currently controls `THERMAL_PROTECTION_BED_HYSTERESIS`.. nothing against, just a reminder.

**Comment by Blue-Marlin** [comment]
There are several ways to check that separately.
On of them is to check for

```
#if WATCH_BED_TEMP_PERIOD > 0
```

instead of

```
#if ENABLED(THERMAL_PROTECTION_BED)
```

**Comment by scalez** [comment]
@thinkyhead @Blue-Marlin I can wrap the heating sanity check code for both the nozzle and the bed with `#if WATCH_TEMP_PERIOD > 0` and `#if WATCH_BED_TEMP_PERIOD > 0`, and set `#define WATCH_BED_TEMP_PERIOD 0` in the PR if we get a consensus. Let me know what should be done and I can commit it. Ultimately, I believe a new macro should be created for `> 0` purposes.

**Comment by jbrazio** [comment]
NVM that will [in the long run] be worst. Let's keep the config flags for the bed as close as possible as the ones for the nozzle.

As this is a safety feature, do we have any idea what would be a good fits-the-most default value ? Let's do some testing if nothing wrong comes then let's ship this enabled by default as we do for nozzle.

**Comment by thinkyhead** [comment]
For "completeness" I suppose it makes sense to wrap the appropriate code in conditionals like the following, so we can disable these "sub-features" by either non-defining or setting the option to zero:

``` cpp
#if ENABLED(THERMAL_PROTECTION_HOTENDS) && WATCH_TEMP_PERIOD > 0
```

We already have some precedent for this, as it allows us to avoid needing to add another "enable" flag.

File: Marlin/example_configurations/tvrrug/Round2/Configuration_adv.h [hunk 1]
```diff
@@ -87,6 +87,18 @@
...
+   * Whenever an M104 or M109 increases the target temperature the firmware will wait for the
+   * WATCH_TEMP_PERIOD to expire, and if the temperature hasn't increased by WATCH_TEMP_INCREASE
+   * degrees, the machine is halted, requiring a hard reset. This test restarts with any M104/M109,
+   * but only if the current temperature is far enough below the target for a reliable test.
+   *
+   * If you get false positives for "Heating failed" increase WATCH_TEMP_PERIOD and/or decrease WATCH_TEMP_INCREASE
+   * WATCH_TEMP_INCREASE should not be below 2.
+   */
+  #define WATCH_TEMP_PERIOD 20                // Seconds
+  #define WATCH_TEMP_INCREASE 2               // Degrees Celsius
```
Comment (thinkyhead) [inline Marlin/example_configurations/tvrrug/Round2/Configuration_adv.h]: These lines should be in the block above, and the `WATCH_BED_TEMP_*` items in the code above should be here.

**Comment by scalez** [comment]
@thinkyhead @jbrazio check out my last two commits, I think it should be good to go now.

**Comment by jbrazio** [comment]
@scalez thanks, it's looking good !

File: Marlin/SanityCheck.h [hunk 2]
```diff
@@ -439,9 +439,9 @@
  */
 #if WATCH_TEMP_PERIOD > 500
   #error WATCH_TEMP_PERIOD now uses seconds instead of milliseconds.
-#elif DISABLED(THERMAL_PROTECTION_HOTENDS) && (defined(WATCH_TEMP_PERIOD) || defined(THERMAL_PROTECTION_PERIOD))
+#elif (DISABLED(THERMAL_PROTECTION_HOTENDS) || WATCH_TEMP_PERIOD <= 0) && (defined(WATCH_TEMP_PERIOD) || (defined(THERMAL_PROTECTION_PERIOD) && WATCH_TEMP_PERIOD > 0))
```
Comment (thinkyhead) [inline Marlin/SanityCheck.h]: The original reason for these tests was to deal with older configurations that didn't have `THERMAL_PROTECTION_HOTENDS` or `THERMAL_PROTECTION_BED` options, so the other options could be enabled without them. This isn't a problem with your new options, which are … new. So you can drop all your changes to `SanityCheck.h` here.

File: Marlin/temperature.cpp [hunk 3]
```diff
@@ -97,14 +97,14 @@ unsigned char soft_pwm_bed;
   int current_raw_filwidth = 0;  //Holds measured filament diameter - one extruder only
 #endif
 
-#if ENABLED(THERMAL_PROTECTION_HOTENDS) || ENABLED(THERMAL_PROTECTION_BED)
+#if (ENABLED(THERMAL_PROTECTION_HOTENDS) && WATCH_TEMP_PERIOD > 0) || (ENABLED(THERMAL_PROTECTION_BED) && WATCH_BED_TEMP_PERIOD > 0)
```
Comment (thinkyhead) [inline Marlin/temperature.cpp]: The `thermal_runaway_protection` function isn't related to `WATCH_TEMP_PERIOD` so you shouldn't be checking it here.

File: Marlin/temperature.cpp [hunk 4]
```diff
@@ -97,14 +97,14 @@ unsigned char soft_pwm_bed;
...
   enum TRState { TRReset, TRInactive, TRFirstHeating, TRStable, TRRunaway };
   void thermal_runaway_protection(TRState* state, millis_t* timer, float temperature, float target_temperature, int heater_id, int period_seconds, int hysteresis_degc);
-  #if ENABLED(THERMAL_PROTECTION_HOTENDS)
+  #if ENABLED(THERMAL_PROTECTION_HOTENDS) && WATCH_TEMP_PERIOD > 0
```
Comment (thinkyhead) [inline Marlin/temperature.cpp]: Same here and below. This condition will break compilation if `WATCH_TEMP_PERIOD` is zero.

File: Marlin/temperature.cpp [hunk 5]
```diff
@@ -200,11 +200,16 @@ static float analog2temp(int raw, uint8_t e);
 static float analog2tempBed(int raw);
 static void updateTemperaturesFromRawValues();
 
-#if ENABLED(THERMAL_PROTECTION_HOTENDS)
+#if ENABLED(THERMAL_PROTECTION_HOTENDS) && WATCH_TEMP_PERIOD > 0
```
Comment (thinkyhead) [inline Marlin/temperature.cpp:203]: This is ok however.

File: Marlin/temperature.cpp [hunk 6]
```diff
@@ -657,14 +662,14 @@ void manage_heater() {
...
-  #if ENABLED(THERMAL_PROTECTION_HOTENDS) || DISABLED(PIDTEMPBED) || HAS_AUTO_FAN
+  #if (ENABLED(THERMAL_PROTECTION_HOTENDS) && WATCH_TEMP_PERIOD > 0) || (ENABLED(THERMAL_PROTECTION_BED) && WATCH_BED_TEMP_PERIOD > 0) || DISABLED(PIDTEMPBED) || HAS_AUTO_FAN
     millis_t ms = millis();
   #endif
 
   // Loop through all extruders
   for (int e = 0; e < EXTRUDERS; e++) {
 
-    #if ENABLED(THERMAL_PROTECTION_HOTENDS)
+    #if ENABLED(THERMAL_PROTECTION_HOTENDS) && WATCH_TEMP_PERIOD > 0
```
Comment (thinkyhead) [inline Marlin/temperature.cpp]: Not here.

File: Marlin/temperature.cpp [hunk 7]
```diff
@@ -728,7 +751,7 @@ void manage_heater() {
 
   #if TEMP_SENSOR_BED != 0
 
-    #if ENABLED(THERMAL_PROTECTION_BED)
+    #if ENABLED(THERMAL_PROTECTION_BED) && WATCH_BED_TEMP_PERIOD > 0
```
Comment (thinkyhead) [inline Marlin/temperature.cpp]: No `WATCH_*_TEMP_PERIOD` check here or below.

File: Marlin/temperature.cpp [hunk 8]
```diff
@@ -1116,7 +1139,23 @@ void tp_init() {
...
+    if (degBed() < degTargetBed() - (WATCH_BED_TEMP_INCREASE + TEMP_BED_HYSTERESIS + 1)) {
+      watch_target_bed_temp = degBed() + WATCH_BED_TEMP_INCREASE;
+      watch_bed_next_ms = millis() + (WATCH_BED_TEMP_PERIOD) * 1000UL;
+    }
+    else
+      watch_bed_next_ms = 0;
+  }
+#endif
+
+#if (ENABLED(THERMAL_PROTECTION_HOTENDS) && WATCH_TEMP_PERIOD > 0) || (ENABLED(THERMAL_PROTECTION_BED) && WATCH_BED_TEMP_PERIOD > 0)
```
Comment (thinkyhead) [inline Marlin/temperature.cpp]: This line should be unchanged too.

**Comment by thinkyhead** [comment]
Other than the needed changes I just listed, it all looks good!

**Comment by scalez** [comment]
@thinkyhead https://github.com/MarlinFirmware/Marlin/pull/3554/commits/abc3bc18afd634fe311801cfbac1fef2ed101dd6 made the needed changes, let me know what you think.

**Comment by thinkyhead** [comment]
@scalez I've squashed all the commits into a single commit and will merge #3653 shortly.

**Pull request closed without merging**